### PR TITLE
Feature: Add Sanitizers to CI

### PR DESCRIPTION
**Pull request by liss-h**

### CODE REVIEW

**Comment by liss-h** [comment]
Not 100% certain that they are enabled now but compiling takes _ages_ now so I would guess they are.

**Review by bigerl — CHANGES_REQUESTED** [review]
Based on the build-times I would enable the sanitizers only on clang-14 static builds. To pass in CXXFLAGs via env is fine here.

File: .github/workflows/reusable_run_test_and_examples_by_linking_type.yml [hunk 1]
```diff
@@ -67,12 +67,12 @@ jobs:
...
         run: |
           cmake -G Ninja -B build_dir -DBUILD_TESTING=ON -DBUILD_EXAMPLES=ON \
-          ${{ fromJSON('["", "-DBUILD_SHARED_LIBS=ON -DCMAKE_POSITION_INDEPENDENT_CODE=ON"]')[inputs.build-shared-libs] }} \
-          ${{ fromJSON('["", "-DCMAKE_CXX_FLAGS=--coverage"]')[inputs.with-coverage] }}  
+          ${{ fromJSON('["", "-DBUILD_SHARED_LIBS=ON -DCMAKE_POSITION_INDEPENDENT_CODE=ON"]')[inputs.build-shared-libs] }}
           cmake --build build_dir
         env:
           CC: ${{ steps.install_cc.outputs.cc }}
           CXX: ${{ steps.install_cc.outputs.cxx }}
+          CXXFLAGS: -fsanitize=address -fsanitize=undefined ${{ fromJSON('["", "--coverage"]')[inputs.with-coverage] }}
```
Comment (bigerl) [inline .github/workflows/reusable_run_test_and_examples_by_linking_type.yml]: [clang suggests](https://clang.llvm.org/docs/AddressSanitizer.html#id3) to use <del>clang++ for linking and</del> (I think that is anyways by invoking the linker through the compiler) to keep function frame pointers

Comment (liss-h) [reply .github/workflows/reusable_run_test_and_examples_by_linking_type.yml]: Added `-fno-omit-frame-pointer -Wno-unused-command-line-argument -fuse-ld=lld`

`-Wno-unused-command-line-argument` is just that clang does complain on every cpp file getting _compiled_ that it doesn't use the linker, which is expected anyways at compile stage.

**Comment by liss-h** [comment]
Ok after rerunning the CI once it appears to work now, no idea why it didn't work the first time

**Review by bigerl — CHANGES_REQUESTED** [review]
Looks good now. Please also add an info "with address and UB sanitizers" to clang-14 static's job name

**Review by bigerl — APPROVED** [review]

**Pull request merged**